tiledb/sm/query/readers/reader_base.cc

```diff
@@ -534,216 +544,127 @@ Status ReaderBase::load_processed_conditions() {
   return Status::Ok();
 }
 
-Status ReaderBase::read_attribute_tiles(
+Status ReaderBase::read_and_unfilter_attribute_tiles(
+    const std::vector<std::string>& names,
+    const std::vector<ResultTile*>& result_tiles) const {
+  auto filtered_data = read_attribute_tiles(names, result_tiles);
```

Comment:
I'm not sure why were are not getting a warning... But if we ever do, I can just call clear on the vector.

And yes, you are following correctly.

Reply:
At very least, the ownership assumptions here need to be really, REALLY well-documented. In BIG BOLD LETTERS.

---

But preferably, I think we can simplify this codepath. We pass `std::vector<ResultTile*>` down the stack, starting here:

https://github.com/TileDB-Inc/TileDB/blob/cbd066a243edd0edf87aecb27a3ffd689ac0b022/tiledb/sm/query/readers/sparse_unordered_with_dups_reader.cc#L233-L235


And AFAICT it ends up here:

https://github.com/TileDB-Inc/TileDB/blob/220877597eeb02cb0c2991de4741cf7e11a1617d/tiledb/sm/query/readers/reader_base.cc#L583-L668

Where we:
- store the reference in `FilteredData`: https://github.com/TileDB-Inc/TileDB/pull/3848/files#diff-b31c25108c53b02d8eac752aee3031ed6665832a2efe7ea049aebfa66a8c62b7R165
     - Why? Where do we use it if `filtered_data` only lives inside this function?
- loop over the tiles and initialize them: https://github.com/TileDB-Inc/TileDB/blob/220877597eeb02cb0c2991de4741cf7e11a1617d/tiledb/sm/query/readers/reader_base.cc#L653-L654

Unless we are using `result_tiles` somewhere else, I think we are really close to being able to pull this out of `::read_tiles` into a data flow that looks like this:
```
auto filtered_data = read_attribute_tiles(names);

// Can we pass this all the way back to `<ReaderType>::dowork()` rather than mutating the vector???
return create_result_tiles(filtered_data);
```

or (slightly less nice, but much better b/c the `std::vector<ResultTiles*>` mutation happens in one place):
```
auto filtered_data = read_attribute_tiles(names);
create_result_tiles(filtered_data, result_tiles);
```
What do you think?